File: competitive-programming/src/edu/algos/li14_linkedlist/easy/L141_LinkedListCycle.java
package edu.algos.li14_linkedlist.easy;

import edu.algos.li00_model.ListNode;
import edu.algos.li14_linkedlist._A01_SingleLinkedList;

public class L141_LinkedListCycle {

	public static void main(String[] args) {
		L141_LinkedListCycle obj = new L141_LinkedListCycle();
		_A01_SingleLinkedList list = new _A01_SingleLinkedList();
		list.insertAtBegining(3);
		list.insertAtEnd(2);
		list.insertAtEnd(0);
		list.insertAtEnd(-4);
		list.head.next.next.next.next = list.head.next;
		System.out.println(obj.hasCycle(list.head));
	}
	
    // Problem: https://leetcode.com/problems/linked-list-cycle/
    // Idea: Use Floyds Cycle Detection algorithm.
    // Use slow and fast ptr's. They can meet at any node.
	public boolean hasCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast){
                return true;
            }
        }
        return false;
    }
}

File: dsa/patterns/09_linkedlist/python/l141_linked_list_cycle.py
"""
Problem: https://leetcode.com/problems/linked-list-cycle/

Idea:
Use Floyds Cycle Detection algorithm.
Use slow and fast ptr's. They can meet at any node.
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def has_cycle(head: ListNode) -> bool:
    slow = fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow is fast:
            return True

    return False


# Example usage
if __name__ == "__main__":
    # Create linked list: 3 -> 2 -> 0 -> -4
    # And make a cycle: -4 -> 2
    n1 = ListNode(3)
    n2 = ListNode(2)
    n3 = ListNode(0)
    n4 = ListNode(-4)

    n1.next = n2
    n2.next = n3
    n3.next = n4
    n4.next = n2   # cycle here

    print(has_cycle(n1))  # True
